sympy/solvers/tests/test_solveset.py

```diff
@@ -1700,6 +1700,29 @@ def test_solve_nonlinear_trans():
     assert nonlinsolve([x**2 - y**2/exp(x)], [x, y]) == soln4
 
 
+def test_issue_19050():
+    # test_issue_19050 --> TypeError removed
+    assert dumeq(nonlinsolve([x + y, sin(y)], [x, y]),
+        FiniteSet((ImageSet(Lambda(n, -2*n*pi), S.Integers), ImageSet(Lambda(n, 2*n*pi), S.Integers)),\
+             (ImageSet(Lambda(n, -2*n*pi - pi), S.Integers), ImageSet(Lambda(n, 2*n*pi + pi), S.Integers))))
+    assert dumeq(nonlinsolve([x + y, sin(y) + cos(y)], [x, y]),
+        FiniteSet((ImageSet(Lambda(n, -2*n*pi - 3*pi/4), S.Integers), ImageSet(Lambda(n, 2*n*pi + 3*pi/4), S.Integers)), \
+            (ImageSet(Lambda(n, -2*n*pi - 7*pi/4), S.Integers), ImageSet(Lambda(n, 2*n*pi + 7*pi/4), S.Integers))))
+
+
+def test_issue_16618():
+    # have to remove the redundancy
+    # AttributeError is removed !
+    eqn = [sin(x)*sin(y), cos(x)*cos(y) - 1]
+    ans = FiniteSet((x, 2*n*pi), (x, 2*n*pi), (2*n*pi, y), (2*n*pi, y), \
+        (x, 2*n*pi + pi), (x, 2*n*pi + pi), (2*n*pi + pi, y), (2*n*pi + pi, y))
+    eq = nonlinsolve(eqn, [x,y])
+
+    for arglen in ans.args:
```

Comment:
```suggestion
    for arglen in eq.args:
```

Reply:
```python
soln = nonlinsolve(eqn, [x, y])
assert soln == ans  # assert soln.dummy_eq(ans)
```

The test as written is not testing that the right result is being obtained, only that it has the right shape.

Reply:
> Yes, since `n` is the dummy so each of them has different indices.

then the place where the solutions are being appended has to be modified to check if a solution that is `dummy_eq`(?) to another is already present.

Reply:
https://github.com/sympy/sympy/blob/566d1e17a32bf3b9df47b9459e277d9a9d0a95ec/sympy/solvers/solveset.py#L3322-L3340

L-3337 is the point where both the real as well as complex solutions get merged. If both of them are same then it gives a redundant solution.

Reply:
> soln = nonlinsolve(eqn, [x, y])
> assert soln == ans  # assert soln.dummy_eq(ans)

I've tried different forms like 
1. assert soln.dummy_eq(ans)
2. assert dumeq(soln, ans)
3. assert soln == ans

But None of them are working, I wonder why this is happening even though the result is correct.

Reply:
> I've tried different forms like

`assert len(sol) == len(ans) and all(a.dummy_eq(b) for i, j in zip(sol, ans) for a,b in zip(i, j))`

Reply:
> the point where both the real as well as complex solutions get merged

```python
    # overall result
    # don't keep duplicate solutions
    filtered_complex = []
    for i in list(new_result_complex):
        for j in list(new_result_real):
            if i.keys() != j.keys():
                continue
            if all(a.dummy_eq(b) for a, b in zip(i.values(),j.values())):
                break
        else:
            filtered_complex.append(i)
    result = new_result_real + filtered_complex]
```

Reply:
But somewhere that the parametric solutions are being added it should be noted that if odd *and* even solutions are being added then they should be collapsed into the general solution `n*pi`  instead of `2*n*pi` and `(2*n + 1)*pi`.

Reply:
It works!

Reply:
Seems ok here:
```python
>>> V = (ax, bx, cx, gx, jx, lx, mx, nx, q) = symbols('ax bx cx gx jx lx mx nx
>>> eq = (ax*q - lx*q - mx, ax - gx*q - lx, bx*q**2 + cx*q - jx*q - nx, q*(-ax lx*q + mx), q*(-ax + gx*q + lx))
>>> nonlinsolve(eq, V)
FiniteSet((lx + mx/q, (-cx*q + jx*q + nx)/q**2, cx, mx/q**2, jx, lx, mx, nx, q), (lx + mx/q, -(cx*q - jx*q - nx)/q**2, cx, mx/q**2, jx, lx, mx, nx, q))
```

Reply:
```
>>> p1 = (lx + mx/q, (-cx*q + jx*q + nx)/q**2, cx, mx/q**2, jx, lx, mx, nx, q)
>>> p2 = (lx + mx/q, -(cx*q - jx*q - nx)/q**2, cx, mx/q**2, jx, lx, mx, nx, q)
>>> p1 == p2
True
```

Reply:
This is correct: the negation results in distribution of the sign automatically.

Reply:
> solutions get merged

This is not the right place to fix it. issue #21022 will fail with only 2 solutions instead of 6

Reply:
The sorting is ok -- there is something non-canonical about the substitution process, however, since with or without the duplicate elimination, sometimes there are 2 solutions and sometimes 6. So could you put in the duplicate removal and then this will be done and we can look at #21022.

Reply:
>  since with or without the duplicate elimination, sometimes there are 2 solutions and sometimes 6


I did not encounter any problem with this. I always get 4 as result. I think you may check with `len(res)`. Even after merging the #21212 in the present PR.

Reply:
The source of the indeterminacy is here:
https://github.com/sympy/sympy/blob/566d1e17a32bf3b9df47b9459e277d9a9d0a95ec/sympy/solvers/solveset.py#L3280

If you replace that with `sol = next(ordered(sol))` there will be 2 solutions to #21022 and if you use `next(reversed(list(ordered(sol))))` there will be 6 solutions. I am not sure why `pop` is being used here, however. It seems to me the handling of `sol` should be in a loop so
```python
sol = set(sol).pop()
...
# solution got for sym
```
becomes
```python
for sol in sol:
    ...
# solution got for sym
```

Reply:
That will give 6 solutions because duplicates obtained are not identified. But if simplification is used, the correct 4 remain. Say `res` is the result, then `FiniteSet(*[tuple([i.simplify() for i in j]) for j in res])` will have a length of 4.

Reply:
I think it will be better to discuss this part in #21223, after this merge. Since IMHO both do not have anything much in common.